File: backend-nodejs/src/utils/dummyData.js
// src/utils/dummyData.js

// Generate random value within range
const randomInRange = (min, max, decimals = 1) => {
  return parseFloat((Math.random() * (max - min) + min).toFixed(decimals));
};

// Generate single sensor reading
export const generateDummyData = () => {
  return {
    timestamp: new Date().toISOString(),
    temperature: randomInRange(18, 32), // °C
    humidity: randomInRange(40, 85),    // %
    soilMoisture: randomInRange(30, 70), // %
    status: 'active',
    greenhouse: {
      id: 'GH-001',
      name: 'ShambaSecure Test Greenhouse',
      location: 'Nairobi, Kenya'
    }
  };
};

// Generate historical data
export const generateHistoricalData = (hours = 24, interval = '1h') => {
  const data = [];
  const intervalMs = interval.endsWith('h') ? parseInt(interval) * 60 * 60 * 1000 : 60 * 60 * 1000;
  const now = Date.now();
  
  // Base values for more realistic trends
  let baseTemp = 25;
  let baseHumidity = 65;
  let baseSoilMoisture = 55;

  for (let i = hours; i >= 0; i--) {
    const timestamp = new Date(now - (i * intervalMs));
    
    // Add slight variations to simulate realistic trends
    baseTemp += randomInRange(-1, 1, 2);
    baseHumidity += randomInRange(-2, 2, 2);
    baseSoilMoisture += randomInRange(-1.5, 1.5, 2);

    // Keep within realistic bounds
    baseTemp = Math.max(18, Math.min(32, baseTemp));
    baseHumidity = Math.max(40, Math.min(85, baseHumidity));
    baseSoilMoisture = Math.max(30, Math.min(70, baseSoilMoisture));

    data.push({
      timestamp: timestamp.toISOString(),
      temperature: parseFloat(baseTemp.toFixed(1)),
      humidity: parseFloat(baseHumidity.toFixed(1)),
      soilMoisture: parseFloat(baseSoilMoisture.toFixed(1))
    });
  }

  return data;
};

// Generate data for specific time ranges
export const generateDataByRange = (range) => {
  const ranges = {
    '1h': { hours: 1, interval: '5m' },
    '6h': { hours: 6, interval: '15m' },
    '24h': { hours: 24, interval: '1h' },
    '7d': { hours: 168, interval: '6h' },
    '30d': { hours: 720, interval: '24h' }
  };

  const config = ranges[range] || ranges['24h'];
  return generateHistoricalData(config.hours, config.interval);
};

File: backend/utils/dummy_data.py
# utils/dummy_data.py
import random
from datetime import datetime, timedelta

def random_in_range(min_val, max_val, decimals=1):
    """Generate random value within range"""
    value = random.uniform(min_val, max_val)
    return round(value, decimals)

def generate_dummy_data():
    """Generate single sensor reading"""
    return {
        'timestamp': datetime.utcnow().isoformat(),
        'temperature': random_in_range(18, 32),  # °C
        'humidity': random_in_range(40, 85),     # %
        'soilMoisture': random_in_range(30, 70), # %
        'status': 'active',
        'greenhouse': {
            'id': 'GH-001',
            'name': 'ShambaSecure Test Greenhouse',
            'location': 'Nairobi, Kenya'
        }
    }

def generate_historical_data(hours=24, interval='1h'):
    """Generate historical sensor data"""
    data = []
    
    # Parse interval
    if interval.endswith('h'):
        interval_minutes = int(interval[:-1]) * 60
    elif interval.endswith('m'):
        interval_minutes = int(interval[:-1])
    else:
        interval_minutes = 60  # Default to 1 hour
    
    # Calculate number of data points
    total_minutes = hours * 60
    num_points = int(total_minutes / interval_minutes) + 1
    
    # Base values for realistic trends
    base_temp = 25.0
    base_humidity = 65.0
    base_soil_moisture = 55.0
    
    current_time = datetime.utcnow()
    
    for i in range(num_points):
        # Calculate timestamp (going backwards from now)
        timestamp = current_time - timedelta(minutes=(num_points - i - 1) * interval_minutes)
        
        # Add slight variations for realistic trends
        base_temp += random_in_range(-1, 1, 2)
        base_humidity += random_in_range(-2, 2, 2)
        base_soil_moisture += random_in_range(-1.5, 1.5, 2)
        
        # Keep within realistic bounds
        base_temp = max(18, min(32, base_temp))
        base_humidity = max(40, min(85, base_humidity))
        base_soil_moisture = max(30, min(70, base_soil_moisture))
        
        data.append({
            'timestamp': timestamp.isoformat(),
            'temperature': round(base_temp, 1),
            'humidity': round(base_humidity, 1),
            'soilMoisture': round(base_soil_moisture, 1)
        })
    
    return data